include/llbuild/Core/BuildDB.h

```diff
@@ -28,22 +28,34 @@ namespace core {
 struct Result;
 class Rule;
 
-class BuildDB {
+/// Delegate interface for use with the build database
+class BuildDBDelegate {
 public:
-  virtual ~BuildDB();
+  virtual ~BuildDBDelegate();
 
   /// Get a unique ID for the given key.
   ///
   /// This method is thread safe.
   ///
   /// \param key [out] The key whose unique ID is being returned.
-  /// \param error_out [out] Error string if return value is 0.
-  virtual KeyID getKeyID(const KeyType& key, std::string *error_out) = 0;
+  virtual KeyID getKeyID(const KeyType& key) = 0;
```

Comment:
I don't think this method is actually used on the delegate.

Reply:
lookupRuleResult uses it

Reply:
That doesn't need to be exposed via the delegate though, that is internal to the DB?

Reply:
It does, in that it is part of the bidirectional mapping.  Since a rule result object has a vector of engine KeyIDs, we need to ask the delegate what the engine key ID is for each key after reading it out of the database.

Reply:
Oh, right, I'm sorry... of course.

More ☕️ needed